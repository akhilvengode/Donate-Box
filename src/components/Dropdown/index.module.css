.dropdown {
  position: relative;
  margin: 1.5rem 0;
}

.dropdown__label {
  font-weight: 500;
}

.dropdown__head {
  box-sizing: border-box;
  display: flex;
  align-items: center;
  justify-content: space-between;
  background-color: white;
  min-height: 50px;
  padding: 1rem;
  width: 100%;
  border: 1px solid lightgrey;
  border-radius: 4px;
  margin-top: 5px;
  cursor: pointer;
}

.dropdown__required {
  color: #be1313;
  margin-left: 4px;
}

.dropdown__error {
  border-color: #be1313;
}

.dropdown__error-message {
  color: #be1313;
  margin-top: 0.4rem;
  font-size: 0.8rem;
  margin: 0;
}

.dropdown__head:focus {
  outline: 2px solid #007fff;
}

.dropdown__placeholder {
  color: #acacac;
  margin: 0;
}

.dropdown__options {
  position: absolute;
  left: 0;
  z-index: 1000;
  background-color: white;
  width: 100%;
  border-radius: 4px;
  margin-top: 0.5rem;
  overflow: hidden;
  box-shadow: 2px 2px 5px lightgrey, -2px -2px 5px lightgrey;
}

.dropdown__option {
  padding: 1rem;
  border-bottom: 1px solid lightgrey;
  cursor: pointer;
}

.dropdown__option:hover {
  background-color: #eeeced;
}

.dropdown__option--active {
  background-color: #ea4c89;
  color: white;
}

.dropdown__option--active:hover {
  background-color: #ea4c89;
}

@media (max-width: 450px) {
  .dropdown__head {
    min-height: 40px;
  }
}
