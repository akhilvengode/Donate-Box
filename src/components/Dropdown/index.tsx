import React, { useEffect, useRef, useState } from "react";
import classnames from "classnames";
import Triangle from "./triangle";
import styles from "./index.module.css";

type OptionType = {
  name: string;
  value: string;
};

type PropType = {
  label: string;
  placeHolder?: string;
  options: Array<OptionType>;
  onChange: (name: string, value: string) => void;
  value?: OptionType;
  required?: boolean;
  error?: boolean;
  errorMessage?: string;
};

const Dropdown: React.FC<PropType> = ({
  label,
  options,
  onChange,
  value,
  placeHolder,
  required = false,
  error = false,
  errorMessage,
}) => {
  const [isOpen, toggle] = useState(false);
  const optionsRef = useRef<HTMLDivElement>(null);
  const dropdownHeadRef = useRef<HTMLDivElement>(null);

  useEffect(() => {
    const toggleOptions = () => {
      toggle(false);
      document.removeEventListener("click", toggleOptions);
    };

    if (isOpen) {
      document.addEventListener("click", toggleOptions);
    }
  }, [isOpen]);

  return (
    <div className={styles.dropdown}>
      <label>
        <span
          onClick={() => {
            dropdownHeadRef.current?.focus();
          }}
          className={styles.dropdown__label}
        >
          {label}
          {required && <span className={styles.dropdown__required}>*</span>}
        </span>
        <div
          id="dropdown-head"
          className={classnames(
            styles.dropdown__head,
            error && styles.dropdown__error
          )}
          ref={dropdownHeadRef}
          tabIndex={0}
          role="button"
          aria-expanded={isOpen}
          onClick={() => toggle((prev) => !prev)}
          onKeyDown={(event) => {
            if (event.key === "Enter") toggle((prev) => !prev);
          }}
        >
          {(value && value.value) || (
              <p className={styles.dropdown__placeholder}>{placeHolder}</p>
            ) ||
            ""}
          <Triangle type={isOpen ? "up" : "down"} />
        </div>
        {error && (
          <p className={styles["dropdown__error-message"]}>{errorMessage}</p>
        )}
      </label>
      {isOpen && (
        <div
          id="dropdown-list"
          ref={optionsRef}
          tabIndex={0}
          role="button"
          className={styles.dropdown__options}
        >
          {options.map(({ name, value: optionValue }) => (
            <div
              className={classnames(
                styles.dropdown__option,
                optionValue === value?.value &&
                  styles["dropdown__option--active"]
              )}
              key={optionValue}
              role="button"
              onClick={() => onChange(name, optionValue)}
            >
              {name}
            </div>
          ))}
        </div>
      )}
    </div>
  );
};

export default Dropdown;
