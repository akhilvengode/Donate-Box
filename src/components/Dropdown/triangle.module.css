.triangle {
  width: 0;
}

.triangle--down,
.triangle--up {
  border-left: 0.31rem solid transparent;
  border-right: 0.31rem solid transparent;
  border-top: 0.62rem solid #b9b9b9;
}

.triangle--up {
  border-bottom: 0.62rem solid #b9b9b9;
  border-top: 0;
}

.triangle--left,
.triangle--right {
  border-top: 0.31rem solid transparent;
  border-bottom: 0.31rem solid transparent;
  border-right: 0.62rem solid #b9b9b9;
}

.triangle--right {
  border-left: 0.62rem solid #b9b9b9;
  border-right: 0;
}
