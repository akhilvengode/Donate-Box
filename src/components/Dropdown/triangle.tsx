import classnames from "classnames";
import styles from "./triangle.module.css";

type PropType = {
  type: "up" | "down" | "right" | "left";
};

const Triangle: React.FC<PropType> = ({ type }) => {
  return (
    <div className={classnames(styles.triangle, styles[`triangle--${type}`])} />
  );
};

export default Triangle;
