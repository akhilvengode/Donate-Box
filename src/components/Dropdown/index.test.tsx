import { shallow } from "enzyme";
import Dropdown from ".";

describe("Testing Dropdown component", () => {
  const mockOptions = [
    { name: "First", value: "first" },
    { name: "Second", value: "second" },
  ];
  const onChangeMock = jest.fn();
  const wrapper = shallow(
    <Dropdown
      options={mockOptions}
      onChange={onChangeMock}
      value={{ name: "Second", value: "second" }}
      label="Test Label"
    />
  );

  it("should match the snapshot", () => {
    expect(wrapper).toMatchSnapshot();
  });

  it("should render the dropdown properly", () => {
    expect(wrapper).toHaveLength(1);
  });

  it("should show/hide when we click on dropdown head", () => {
    expect(wrapper.find("#dropdown-list")).toHaveLength(0);
    wrapper.find("#dropdown-head").simulate("click");
    expect(wrapper.find("#dropdown-list")).toHaveLength(1);
    wrapper.find("#dropdown-head").simulate("click");
    expect(wrapper.find("#dropdown-list")).toHaveLength(0);
  });

  it("should call onChange when change event happens", () => {
    wrapper.find("#dropdown-head").simulate("click");
    wrapper.find(".dropdown__option").at(0).simulate("click");
    expect(onChangeMock).toBeCalledTimes(1);
  });
});
