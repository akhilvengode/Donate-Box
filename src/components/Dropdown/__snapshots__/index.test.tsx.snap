// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing Dropdown component should match the snapshot 1`] = `
<div
  className="dropdown"
>
  <label>
    <span
      className="dropdown__label"
      onClick={[Function]}
    >
      Test Label
    </span>
    <div
      aria-expanded={false}
      className="dropdown__head"
      id="dropdown-head"
      onClick={[Function]}
      onKeyDown={[Function]}
      role="button"
      tabIndex={0}
    >
      second
      <Triangle
        type="down"
      />
    </div>
  </label>
</div>
`;
